### PR TITLE
chore: allow updating har while routing

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/server/har/harTracer.ts [hunk 1]
```diff
@@ -252,10 +256,12 @@ export class HarTracer {
   private async _onRequestFinished(request: network.Request, response: network.Response | null) {
     if (!response)
       return;
-    const page = request.frame()._page;
     const harEntry = this._entryForRequest(request);
     if (!harEntry)
       return;
+    const page = request.frame()._page;
+    if (this._page && page !== this._page)
```
Comment (dgozman) [inline packages/playwright-core/src/server/har/harTracer.ts:263]: This check is probably not needed, given the one in `_onRequest`.

File: packages/playwright-core/src/server/har/harTracer.ts [hunk 2]
```diff
@@ -108,7 +110,7 @@ export class HarTracer {
     return (request as any)[this._entrySymbol];
   }
 
-  private _ensurePageEntry(page: Page): har.Page | undefined {
+  private _createPageEntryIfNeeded(page: Page): har.Page | undefined {
```
Comment (dgozman) [inline packages/playwright-core/src/server/har/harTracer.ts:113]: Let's add a `page !== this._page` check in this method to be extra careful.

**Review by dgozman — APPROVED** [review]

**Review by yury-s — APPROVED** [review]
As long as we are ok with 2 ways to record har the change looks good to me.

**Review by rwoll — COMMENTED** [review]
LGTM, assuming update option as a `bool` is enough. I expected something like:

- `"always"`: create a new HAR from or overwrite the existing one
- `"never"`: use existing HAR
- `"on-missing"`: only update if it's missing on disk

However, maybe this is better handled at the @playwright/test level.

**Pull request merged**